src/evo/cbtx.cpp

```diff
@@ -314,8 +322,139 @@ bool CalcCbTxMerkleRootQuorums(const CBlock& block, const CBlockIndex* pindexPre
+        // Previous block Coinbase contains a non-null CL: We must insert the same sig or a better (newest) one
+        if (chainlock_handler.GetBestChainLock().IsNull()) {
+            // We don't know any CL, therefore inserting the CL of the previous block
+            bestCLHeightDiff = prevBlockCoinbaseChainlock->second + 1;
+            bestCLSignature = prevBlockCoinbaseChainlock->first;
+            return true;
+        }
+
+        // We check if our best CL is newer than the one from previous block Coinbase
+        auto curCLHeight = chainlock_handler.GetBestChainLock().getHeight();
+        auto prevCLHeight = (static_cast<uint32_t>(nHeight - 1) - prevBlockCoinbaseChainlock->second - 1);
+        if (curCLHeight < prevCLHeight) {
+            // Our best CL isn't newer: inserting CL from previous block
+            bestCLHeightDiff = prevBlockCoinbaseChainlock->second + 1;
+            bestCLSignature = prevBlockCoinbaseChainlock->first;
+        }
+        else {
+            // Our best CL is newer
+            bestCLHeightDiff = (static_cast<uint32_t>(nHeight) - chainlock_handler.GetBestChainLock().getHeight() - 1);
+            bestCLSignature = chainlock_handler.GetBestChainLock().getSig();
+        }
+
+        return true;
+    }
+    else {
+        // Previous block Coinbase has no CL. We can either insert null or any valid CL
+        if (chainlock_handler.GetBestChainLock().IsNull()) {
+            // We don't know any CL, therefore inserting a null CL
+            bestCLHeightDiff = 0;
+            bestCLSignature.Reset();
+            return false;
+        }
+
+        // Inserting our best CL
+        bestCLHeightDiff = (static_cast<uint32_t>(nHeight) - chainlock_handler.GetBestChainLock().getHeight() - 1);
+        bestCLSignature = chainlock_handler.GetBestChainLock().getSig();
+
+        return true;
+    }
+}
+
+
 std::string CCbTx::ToString() const
 {
-    return strprintf("CCbTx(nVersion=%d, nHeight=%d, merkleRootMNList=%s, merkleRootQuorums=%s)",
-        nVersion, nHeight, merkleRootMNList.ToString(), merkleRootQuorums.ToString());
+    return strprintf("CCbTx(nVersion=%d, nHeight=%d, merkleRootMNList=%s, merkleRootQuorums=%s, bestCLHeightDiff=%d, bestCLSig=%s)",
+        nVersion, nHeight, merkleRootMNList.ToString(), merkleRootQuorums.ToString(), bestCLHeightDiff, bestCLSignature.ToString());
+}
+
+std::optional<CCbTx> GetCoinbaseTx(const CBlockIndex* pindex)
+{
+    if (pindex == nullptr) {
+        return std::nullopt;
+    }
+
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
+        return std::nullopt;
+    }
+
+    CTransactionRef cbTx = block.vtx[0];
+    CCbTx cbTxPayload;
+    if (!GetTxPayload(*cbTx, cbTxPayload)) {
+        return std::nullopt;
+    }
+
+    return cbTxPayload;
+}
+
+std::optional<std::pair<CBLSSignature, uint32_t>> GetNonNullCoinbaseChainlock(const CBlockIndex* pindex)
+{
+    auto opt_cbtx = GetCoinbaseTx(pindex);
```

Comment:
we should probably store`cbtx.bestCLSignature, cbtx.bestCLHeightDiff` in state somewhere so we don't have to load blocks from disk

Reply:
Do you mean to cache `{cbtx.bestCLSignature, cbtx.bestCLHeightDiff}` per block ?

Reply:
Implemented in #5347 by the way